### PR TITLE
Support generating a DOT file for subtask graph

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Add an optional config `dump_subtask_graph` when submitting graphs, it will generate the complete chunk graph, those chunks be colored with same color will be united as a subgraph in DOT, and then connect subgraphs according to the subtask dependences. Specially, if one task has multiple stages, the edge's style will be "dotted" if the connected subtasks belong to different stages.

Here are some examples:

- Example1
```
In [12]: a = mt.random.rand(10, 10, chunk_size=3)

In [13]: b = a + 1 - 2

In [14]: c = b.sum(axis=1)

In [15]: c.execute(extra_config={"dump_subtask_graph": True})
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100.0/100 [00:00<00:00, 360.90it/s]
Out[15]:
array([-4.79605122, -4.17584986, -3.62345371, -6.74078572, -5.11383525,
       -6.32972501, -4.45877359, -6.85511891, -4.79619598, -4.21884735])
```
image

- example2
```
In [3]: a = mt.random.rand(10, 10, chunk_size=3)

In [4]: df = md.DataFrame(a)

In [5]: df[df > 0.5].sort_values(by=1).execute(extra_config={"dump_subtask_graph": True})
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100.0/100 [00:00<00:00, 148.39it/s]
Out[5]:
          0         1         2         3         4         5         6         7         8         9
8       NaN  0.559153  0.991407       NaN  0.579143  0.592087       NaN       NaN  0.731116  0.842489
2  0.676814  0.803772  0.881030       NaN       NaN       NaN       NaN       NaN       NaN       NaN
9       NaN  0.892045  0.537926       NaN  0.760115  0.729152       NaN  0.558372  0.810266       NaN
7  0.868231  0.910737       NaN       NaN  0.720215  0.635155  0.588498  0.972826  0.558915       NaN
0       NaN       NaN  0.810465  0.609881       NaN       NaN  0.950973       NaN       NaN  0.610496
1       NaN       NaN       NaN  0.876199  0.903854       NaN  0.985229       NaN  0.779864       NaN
3       NaN       NaN       NaN       NaN  0.946391  0.652453       NaN       NaN  0.938840  0.538927
4  0.850098       NaN  0.872272  0.882686  0.667572       NaN       NaN       NaN       NaN  0.535971
5       NaN       NaN  0.691697       NaN  0.857972  0.808739  0.522473       NaN       NaN       NaN
6       NaN       NaN       NaN       NaN       NaN       NaN  0.737246  0.565483  0.914582       NaN
```
image

- Example3 (has two stages)
``` Python
In [17]: rs = np.random.RandomState(0)
    ...: raw = pd.DataFrame(
    ...:     {
    ...:         "c1": rs.randint(20, size=100),
    ...:         "c2": rs.choice(["a", "b", "c"], (100,)),
    ...:         "c3": rs.rand(100),
    ...:     }
    ...: )
    ...: mdf = md.DataFrame(raw, chunk_size=20)

In [18]: r = mdf.groupby("c2").agg("sum")

In [19]: r.execute(extra_config={"dump_subtask_graph": True})
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100.0/100 [00:00<00:00, 288.89it/s]
Out[19]:
     c1         c3
c2
a   305  16.312997
b   254  14.502106
c   329  18.307300
```
image

## Related issue number

Fixes #xxxx

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

**Comment by wjsi** [comment]
Will this be generated and stored at specific locations when some options (for inst., an environment variable) are set? Web API and private session functions also seems useful.

**Comment by hekaisheng** [comment]
> Will this be generated and stored at specific locations when some options (for inst., an environment variable) are set? Web API and private session functions also seems useful.

Haven't been implemented in this PR.

**Comment by hekaisheng** [comment]
Use different colors if edges are cross stages, new graph for example3:

image

**Review by qinxuye — DISMISSED** [review]
LGTM

File: mars/services/task/supervisor/processor.py [hunk 1]
```diff
@@ -51,6 +53,8 @@
 
 logger = logging.getLogger(__name__)
 
+MARS_ENABLE_DUMPING_SUBTASK_GRAPH = bool(os.environ.get("MARS_DUMP_SUBTASK_GRAPH", 0))
```
Comment (wjsi) [inline mars/services/task/supervisor/processor.py:56]:
```suggestion
MARS_ENABLE_DUMPING_SUBTASK_GRAPH = int(os.environ.get("MARS_DUMP_SUBTASK_GRAPH", 0))
```

File: mars/services/task/supervisor/processor.py [hunk 2]
```diff
@@ -452,8 +462,31 @@ def gen_result(self):
...
+        from .graph_visualizer import GraphVisualizer
+
+        try:  # pragma: no cover
+            import graphviz
+        except ImportError:
+            graphviz = None
+
+        dot = GraphVisualizer(self).to_dot()
+        directory = tempfile.gettempdir()
+        file_name = f"mars-{self.task_id}"
```
Comment (wjsi) [inline mars/services/task/supervisor/processor.py:475]: Contain `stage_id` in dumped file name.

File: mars/services/task/supervisor/graph_visualizer.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                )
+                for c in subtask.chunk_graph.results:
+                    result_chunk_to_subtask[c.key] = [current_stage, current_cluster]
+                subgraph_index += 1
+            current_stage += 1
+        sio.write("}")
+        return sio.getvalue()
+
+    @classmethod
+    def _export_subtask_to_dot(
```
Comment (wjsi) [inline mars/services/task/supervisor/graph_visualizer.py:66]: Is it possible to display subtask id in dag?

**Comment by hekaisheng** [comment]
Display subtask id in each subgraph, now graph looks as below:
image

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**